### PR TITLE
PLT-321: Implement makeAutoBalancedTransactionWithUtxoProvider

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Formatting, PNG optimization, etc. are updated
    - [x] Important changes are reflected in changelog.d of the affected packages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [x] Reviewer requested

### CODE REVIEW

**Review by koslambrou — COMMENTED** [review]
* Why move coin selection in `Ledger.Fee`? Isn't our coin selection algorithm specific to the emulator? In the future, I think coin selection should be done by a separate component (not by cardano-api or cardano-ledger-api) as it is complicated logic. 

* On a related note, IMO, I think there is some value for having logs for balancing in the emulator, which helps debugging. Unless you had a specific reason to remove them?

**Review by berewt — APPROVED** [review]
LGTM, one minor change.

File: plutus-ledger/src/Ledger/Fee.hs [hunk 1]
```diff
@@ -74,6 +108,232 @@ makeAutoBalancedTransaction params utxo (CardanoBuildTx txBodyContent) pChangeAd
...
+    let returnCollateral = Tx.getTxBodyContentReturnCollateral txWithinputsAdded
+
+    if Value.isZero (fold collateral)
+        && null (C.collectTxBodyScriptWitnesses txWithinputsAdded) -- every script has a redeemer, no redeemers -> no scripts
+        && null returnCollateral then
+        -- Don't add collateral if there are no plutus scripts that can fail
+        -- and there are no collateral inputs or outputs already
+        pure txWithinputsAdded
+    else do
+        let collAddr = maybe cChangeAddr (\(Tx.TxOut (C.TxOut aie _tov _tod _rs)) -> aie) returnCollateral
```
[the remark this answers is not in the document]
Comment (sjoerdvisscher) [reply plutus-ledger/src/Ledger/Fee.hs:214]: No, I need the cardano address, not the plutus address. You can fix this in your PR when you convert our use of Address to the cardano type.

[the remark this answers is not in the document]
Comment (berewt) [reply plutus-ledger/src/Ledger/Fee.hs:214]: oh right, I was confused because of my current work.

**Comment by sjoerdvisscher** [comment]
> I think there is some value for having logs for balancing in the emulator, which helps debugging. Unless you had a specific reason to remove them?

@koslambrou We still log the transaction before and after balancing, so this information can easily be derived. And I have never seen the output of these logging statements. Also the functions in `Ledger.Fee` don't use an effect system, so adding logging would be somewhat messy.

**Comment by sjoerdvisscher** [comment]
> Isn't our coin selection algorithm specific to the emulator? 
@koslambrou Why do you think that? I don't think there's anything emulator-specific in there. But otherwise good point, coin selection shouldn't be fixed. So I'll change the wallet utxo argument to a function that returns utxos given a value.

**Comment by koslambrou** [comment]
@koslambrou Why do you think that? I don't think there's anything emulator-specific in there. But otherwise good point, coin selection shouldn't be fixed. So I'll change the wallet utxo argument to a function that returns utxos given a value.

What I meant was that users would typically decide what the coin selection algorithm should be right? They can use whatever algorithm they want. The one we provide in the emulator should maybe not be the one used in production? Someone might want to use the one from cardano-wallet, no?

File: plutus-ledger/src/Ledger/Fee.hs [hunk 2]
```diff
@@ -74,6 +109,245 @@ makeAutoBalancedTransaction params utxo (CardanoBuildTx txBodyContent) pChangeAd
       Nothing
 
 
+-- | Creates a balanced transaction by calculating the execution units, the fees and then the balance.
+-- If the balance is negative the utxo provider is asked to pick extra inputs to make the balance is positive,
+-- which is then assigned to the change address.
+-- The collateral is similarly balanced.
+-- Unlike `makeAutoBalancedTransaction` this function also balances non-Ada.
+makeAutoBalancedTransactionWithUtxoProvider
```
Comment (koslambrou) [inline plutus-ledger/src/Ledger/Fee.hs:117]: Should we delete `makeAutoBalancedTransaction` if this one is more complete and balances non-Ada?

**Review by koslambrou — APPROVED** [review]

File: plutus-ledger/src/Ledger/Fee.hs (hunk 2 above)
Comment (sjoerdvisscher) [reply plutus-ledger/src/Ledger/Fee.hs:117]: I can imagine that you would not want automatic non-Ada balancing. It could hide bugs.

**Pull request merged**